oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/parser/ISegmentParserListenerManager.java

```diff
@@ -20,6 +20,10 @@
 
 import org.apache.skywalking.oap.server.analyzer.provider.trace.parser.listener.AnalysisListenerFactory;
 
+import java.util.List;
+
 public interface ISegmentParserListenerManager {
     void add(AnalysisListenerFactory analysisListenerFactory);
+
+    List<AnalysisListenerFactory> getSpanListenerFactories();
```

Comment:
I'm not entirely sure what you made about the `listenerManagers`. 

I added `getSpanListenerFactories` so I can make sure my `AnalysisListenerFactory` was executed first. Calling `ISegmentParserListenerManager.add()` would make it the last factory called.

Reply:
In those fields, they are hard coded as `SegmentParserListenerManager`, rather than `ISegmentParserListenerManager`. I am not sure whether this is your point.

Reply:
Ah, I think I get what you mean. No, I don't plan on replacing the implementation. I just want to be able to make sure my `AnalysisListenerFactory` is the first called during `TraceAnalyzer.doAnalysis()`. I can do this by calling `getSpanListenerFactories` and doing `getSpanListenerFactories().add(0, myListener)`.

I also noticed that `EventAnalyzerListenerFactoryManager.factories()` exists. Should I also rename that to `IEventAnalyzerListenerFactoryManager.getEventAnalyzerListenerFactories()`? That would make it consistent with `ISegmentParserListenerManager` and  `ILogAnalysisListenerFactoryManager`.

Reply:
I think you should simply do `(SegmentParserListenerManager)manager`.`getSpanListenerFactories().add(0, myListener)`, directly?

I think we don't intend to expose the list because it would be tricky to control the analyzer sequence.

Reply:
`SegmentParserListenerManager.getSpanListenerFactories` is not public though.

And as I noted, both `EventAnalyzerListenerFactoryManager` & `ILogAnalysisListenerFactoryManager` expose their factory lists. Should they not expose them either?

Reply:
OK. This would be only for your fork :) For upstream, either is fine.